hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/update/server/SCMUpdateClientManager.java

```diff
@@ -0,0 +1,149 @@
+    UUID clientId = UUID.randomUUID();
+    int retryCount = 5;
+    while (clients.containsKey(clientId)) {
+      if (retryCount > 0) {
+        clientId = UUID.randomUUID();
+        retryCount--;
+      } else {
+        throw new SCMException("Failed to add CRL client with random clientId" +
+            " collision", SCMException.ResultCodes.FAILED_TO_ADD_CRL_CLIENT);
+      }
+    }
+
+    SCMUpdateClientInfo clientInfo = new SCMUpdateClientInfo(clientId);
+    clients.put(clientId, clientInfo);
+    return clientId;
+  }
+
+  // this does not necessarily produce a server response via responseObserver.
+  public void handleClientUpdate(UpdateRequest request,
+      StreamObserver<UpdateResponse> responseObserver) {
+    UUID clientId = SCMUpdateClientInfo.fromClientIdProto(
+        request.getClientId());
+
+    // Unknown client update
+    if (!clients.containsKey(clientId)) {
+      responseObserver.onError(Status.INVALID_ARGUMENT
+          .withDescription("Client must subscribe before it can " +
+              "send/receive updates")
+          .asException());
+    }
+
+    // record the server to client channel
+    SCMUpdateClientInfo clientInfo = clients.get(clientId);
+    if (clientInfo.getResponseObserver() == null) {
+      clientInfo.setResponseObserver(responseObserver);
+    }
+
+    if (handlers.containsKey(request.getUpdateType())) {
+      handlers.get(request.getUpdateType())
+          .handleClientRequest(request, clientInfo);
+    } else {
+      responseObserver.onError(Status.INVALID_ARGUMENT
+          .withDescription("Unknown client update type.")
+          .asException());
+    }
+  }
+
+  /**
+   * Remove client by client Id.
+   * @param clientId - client Id
+   * @return true if client is removed, false otherwise.
+   */
+  public boolean removeClient(UUID clientId) {
+    if (clients.containsKey(clientId)) {
+      SCMUpdateClientInfo clientInfo = clients.remove(clientId);
+      handlers.values().forEach(handler -> handler.onRemoveClient(clientInfo));
+      LOG.info("Client {} removed.", clientId);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Remove client by its responseObserver obj.
+   * @param responseObserver - response observer of the client
+   * @return true if client is removed, false otherwise.
+   */
+  public boolean removeClient(StreamObserver<UpdateResponse> responseObserver) {
+    UUID clientId = null;
+    for (SCMUpdateClientInfo client : clients.values()) {
+      if (client.getResponseObserver() == responseObserver) {
+        clientId = client.getClientId();
+        break;
+      }
+    }
+    if (clientId != null) {
+      LOG.debug("Remove client {} by responseObserver", clientId);
+      removeClient(clientId);
+      return true;
+    }
+    LOG.debug("Remove client {} by responseObserver not found!");
+    return false;
+  }
+
+  public void onUpdate(Type type) {
+    if (handlers.containsKey(type)) {
+      handlers.get(type).onUpdate();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **142 and 143**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **53**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).